File: app/src/main/java/io/mochadwi/mobilenews/news_source/NewsSourcePresenter.java
package io.mochadwi.mobilenews.news_source;

import io.mochadwi.mobilenews.news_source.model.NewsSourceModel;
import io.mochadwi.mobilenews.network.RESTClient;
import io.mochadwi.mobilenews.network.RESTGenerator;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;
import retrofit2.Retrofit;

public class NewsSourcePresenter implements NewsSourceContract.Presenter {

    private final NewsSourceContract.View mView;
    private final Retrofit mRest = RESTGenerator.Companion.getRetrofitClient();

    public NewsSourcePresenter(NewsSourceContract.View view) {
        this.mView = view;
    }

    @Override
    public void start() {
        mView.setPresenter(this);
    }

    @Override
    public void getNews(String apiKey) {
        mView.showProgress();

        mRest
                .create(RESTClient.class)
                .getRecommendedSources(apiKey)
                .enqueue(new Callback<NewsSourceModel>() {
                    @Override
                    public void onResponse(Call<NewsSourceModel> call, Response<NewsSourceModel> response) {
                        mView.hideProgress();
//                        mView.showToast(response.body().toString());
                        mView.setRecyclerView(response.body());
                    }

                    @Override
                    public void onFailure(Call<NewsSourceModel> call, Throwable t) {
                        mView.hideProgress();
                        mView.showToast(t.getMessage());
                    }
                });
    }
}

File: app/src/dev/java/io/mochadwi/mobilenews/dev/news_source/NewsSourcePresenter.kt
package io.mochadwi.mobilenews.dev.news_source

import io.mochadwi.mobilenews.network.RESTClient
import io.mochadwi.mobilenews.network.RESTGenerator
import io.mochadwi.mobilenews.dev.news_source.model.NewsSourceModel
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * Created by mochadwi on 3/13/18.
 */
class NewsSourcePresenter(private val mView: NewsSourceContract.View) : NewsSourceContract.Presenter {
    private val mRest = RESTGenerator.getRetrofitClient()

    override fun start() {
        mView.setPresenter(this)
    }

    override fun getNews(apiKey: String) {
        mView.showProgress()

        mRest
                .create(RESTClient::class.java)
                .getRecommendedSources(apiKey)
                .enqueue(object : Callback<NewsSourceModel> {
                    override fun onResponse(call: Call<NewsSourceModel>, response: Response<NewsSourceModel>) {
                        mView.hideProgress()
//                        mView.showToast(response.body().toString());
                        mView.setRecyclerView(response.body()!!)
                    }

                    override fun onFailure(call: Call<NewsSourceModel>, t: Throwable) {
                        mView.hideProgress()
                        mView.showToast(t.message!!)
                    }
                })
    }
}